src/Analyzers/Core/CodeFixes/FileHeaders/AbstractFileHeaderCodeFixProvider.cs

```diff
@@ -114,19 +108,26 @@ private static SyntaxNode ReplaceHeader(ISyntaxFacts syntaxFacts, AbstractFileHe
             // followed by a comment which is considered a header comment.
             var possibleLeadingSpaces = string.Empty;
 
+            // Track the first code comment. Any comments found later should not be removed.
+            // 0: Init
+            // 1: First comment found
+            // 2: NewLine ending the first comment found
+            var firstCommentState = 0;
```

Comment:
yikes.  this is getting *super* complicated.  Does this layer have access to C:\github\roslyn\src\Workspaces\SharedUtilitiesAndExtensions\Compiler\Core\Utilities\Matcher.cs ?

If so, tehre is a much easier way to do this, effectively by building up 'rich structural regexes' like so:

```c#
            var commentLine = Matcher.Sequence(whitespace, anyCommentMatcher, whitespace, endOfLine);

            _oneOrMoreBlankLines = Matcher.OneOrMore(singleBlankLine);
            _bannerMatcher =
                Matcher.Sequence(
                    Matcher.OneOrMore(commentLine),
                    _oneOrMoreBlankLines);
            _fileBannerMatcher =
                Matcher.Sequence(
                    Matcher.OneOrMore(commentLine),
                    Matcher.Repeat(singleBlankLine));
```

These are basically 'matchers' that can run over a sequence of objects, not a sequence of characters.  You could use this (like above) to much more easily express what you're looking for with the sequence of trivia elements.

Reply:
see: https://github.com/dotnet/roslyn/blob/bd7a354037987e6895b80989d4adabba5860fda2/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/SyntaxFacts/AbstractSyntaxFacts.cs#L64-L72 and https://github.com/dotnet/roslyn/blob/bd7a354037987e6895b80989d4adabba5860fda2/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Services/SyntaxFacts/AbstractSyntaxFacts.cs#L331-L340

Reply:
basically, instead of manually unrolling a dfa and tracking states yourself, express it much more simply with a matching structure that you can see if it matches or not with.

Reply:
> yikes

Exactly my thought when I implemented it.

I implemented it by using `ISyntaxFacts.GetFileBanner` and remove any trivia considered part of the banner from `root.GetLeadingTrivia()`. See ca6d52d1 for details. This changed the behavior of the fixer in three aspects:

* Broken comments are now kept (This one is interesting: I changed the behavior in this PR, so broken comments were replaced. With this change, we are back to the previous behavior: broken comments are kept)
* `ISyntaxFacts.GetFileBanner` expects the banner to be at the top of the file (no leading whitespace or newlines are allowed)
* There was a calculation of `leadingSpaces` present which seemed to calculate whitespace to add before single line comments. This *feature* was not under test. I removed the calculation in ca6d52d1 (before: `var newHeaderTrivia = CreateNewHeader(syntaxFacts, leadingSpaces + fileHeaderHelper.CommentPrefix, expectedFileHeader, newLineTrivia.ToFullString());` after `var newHeaderTrivia = CreateNewHeader(syntaxFacts, fileHeaderHelper.CommentPrefix, expectedFileHeader, newLineTrivia.ToFullString());`)

All the tests that changed with the usage of `ISyntaxFacts.GetFileBanner` are in commit 0ac8ec78.

I think we can keep the logic with `ISyntaxFacts.GetFileBanner` as is. Should we change the `ISyntaxFacts.GetFileBanner` implementations to accept leading whitespace/newlines?

If this is the right way forward I may need to refactor the `existingBannerIndices` calculation and we should also adopt the logic (re-use `ISyntaxFacts.GetFileBanner`)  here:

https://github.com/dotnet/roslyn/blob/8d278c269befa4a2bad29cc001fcb756b54341b2/src/Analyzers/Core/Analyzers/FileHeaders/AbstractFileHeaderHelper.cs#L41-L135